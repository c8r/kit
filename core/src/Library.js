import React from 'react'
import PropTypes from 'prop-types'
import nano from 'nano-style'
import {
  BrowserRouter,
  StaticRouter,
  Route,
  NavLink,
  Link,
  withRouter
} from 'react-router-dom'
import { Grid, Box } from './ui'
import Frame from './Frame'
import Catch from './Catch'

const Root = nano('div')({
  display: 'flex',
  alignItems: 'flex-start',
  height: '100vh'
})

const Main = nano('div')({
  flex: '1 1 auto',
  height: '100vh',
  overflowY: 'auto',
  WebkitOverflowScrolling: 'touch'
})

const Card = nano(Link)({
  display: 'block',
  color: 'inherit',
  textDecoration: 'none',
  overflow: 'hidden',
  border: '1px solid #f6f6f6',
  '&:hover': {
    borderColor: '#ddd'
  }
})

const SideBar = nano('div')({
  width: '192px',
  flex: 'none',
  height: '100vh',
  overflowY: 'auto',
  WebkitOverflowScrolling: 'touch',
  borderRight: '1px solid #f6f6f6'
})

const NavItem = nano(NavLink)({
  display: 'block',
  paddingLeft: '8px',
  paddingRight: '8px',
  paddingTop: '4px',
  paddingBottom: '4px',
  fontSize: '12px',
  fontWeight: 'bold',
  textDecoration: 'none',
  color: 'inherit',
  WebkitUserSelect: 'none',
  userSelect: 'none',
  '&.active': {
    color: 'white',
    backgroundColor: 'black'
  }
})

const Router = typeof document !== 'undefined' ? BrowserRouter : StaticRouter

export class Library extends React.Component {
  render() {
    const { basename, ...props } = this.props

    return (
      <Router basename={basename} context={{}}>
        <LibraryApp {...props} />
      </Router>
    )
  }
}

const LibraryApp = withRouter(
  class extends React.Component {
    static propTypes = {
      title: PropTypes.string,
      examples: PropTypes.array,
      renderSideNav: PropTypes.func,
      renderCard: PropTypes.func,
      useFrame: PropTypes.bool,
    }

    getExampleChildren = ({ children }) =>
      React.Children.toArray(children)
        .filter(c => c.type._kitLibraryExample)
        .filter(c => !!c.props.name)
        .map(c => ({
          name: c.props.name,
          element: c.props.children
        }))

    getHead = ({ children }) => {
      const [ head ] = React.Children.toArray(children)
        .filter(c => c.type === Head)
      if (!head || !head.props) return null
      return head.props.children
    }

    render() {
      const {
        title,
        renderSideNav,
        renderCard,
      } = this.props

      const examples =
        this.props.examples || this.getExampleChildren(this.props)
      const head = this.getHead(this.props)

      const sidenav =
        typeof renderSideNav === 'function' ? (
          renderSideNav({ ...this.state, title, examples })
        ) : (
          <SideNav title={title} examples={examples} />
        )

      return (
        <Root>
          <SideBar>{sidenav}</SideBar>
          <Main>
            <Route
              exact
              path="/"
              render={() => (
                <Grid>
                  {examples.map(example => (
                    typeof renderCard === 'function' ? (
                      renderCard({ example, Card, Link })
                    ) : (
                      <Card key={example.name} to={'/' + example.name}>
                        <Catch>
                          <Box p={2}>
                            <ExampleFrame
                              head={head}
                              example={example}
                            />
                          </Box>
                        </Catch>
                      </Card>
                    )
                  ))}
                </Grid>
              )}
            />
            {examples.map(example => (
              <Route
                key={example.name}
                path={'/' + example.name}
                render={() => (
                  <Box>
                    <Catch>
                      <ExampleFrame
                        head={head}
                        example={example}
                        height='100vh'
                      />
                    </Catch>
                  </Box>
                )}
              />
            ))}
          </Main>
        </Root>
      )
    }
  }
)

const ExampleFrame = ({
  example,
  useFrame,
  head,
}) => {
  if (!useFrame) {
    return example.element
  }
  return (
    <Frame head={head} {...props}>
      {example.element}
    </Frame>
  )
}

class SideNav extends React.Component {
  static propTypes = {
    title: PropTypes.string,
    examples: PropTypes.array
  }

  static defaultProps = {
    title: 'Kit',
    examples: []
  }

  render() {
    const { examples, title } = this.props

    return (
      <React.Fragment>
        <NavItem exact to="/">
          {title}
        </NavItem>
        {examples.map(example => (
          <NavItem key={example.name} to={'/' + example.name}>
            {example.name}
          </NavItem>
        ))}
      </React.Fragment>
    )
  }
}

export class Example extends React.Component {
  static _kitLibraryExample = true
  static propTypes = {
    name: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  }

  render() {
    const { children } = this.props
    return <React.Fragment>{children}</React.Fragment>
  }
}

export const Detail = withRouter(
  class extends React.Component {
    render() {
      const { location } = this.props
      if (location.pathname === '/') return false

      return <div {...this.props} />
    }
  }
)

export class Head extends React.Component {
  render () {
    return false
  }
}

export default Library
