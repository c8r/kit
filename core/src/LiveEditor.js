import React, { Component } from 'react'
import PropTypes from 'prop-types'
import matter from 'gray-matter'
import nano from 'nano-style'

import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

import Catch from './Catch'
import { Box } from './ui'

const Editor = nano(LiveEditor)({
  fontFamily: 'Menlo, monospace',
  fontSize: '14px',
  padding: '8px',
  backgroundColor: '#f6f6f6',
  outline: 'none'
})

class KitEditor extends Component {
  static propTypes = {
    code: PropTypes.string.isRequired,
    scope: PropTypes.object,
    theme: PropTypes.object
  }

  render() {
    const { code, scope, theme = {} } = this.props
    const { content, data = {} } = matter(code)

    return (
      <Catch>
        <LiveProvider
          mountStylesheet={false}
          scope={{
            theme,
            ...scope
          }}
          code={content}
          transformCode={code => `<div>${code}</div>`}
        >
          <Box>
            <LivePreview />
            <Box mt={2}>
              <Editor />
            </Box>
            <Box w={1}>
              <LiveError
                style={{
                  backgroundColor: 'red',
                  color: 'white',
                  padding: '.5rem',
                  height: 'auto'
                }}
              />
            </Box>
          </Box>
        </LiveProvider>
      </Catch>
    )
  }
}

export default KitEditor
