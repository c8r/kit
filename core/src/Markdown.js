import React from 'react'
import { transform } from 'buble'
import mdx from '@mdx-js/mdx'
import { MDXTag } from '@mdx-js/tag'

export default ({ children, components, theme = {} }) => {
  const jsx = mdx.sync(children).replace('export default ({components}) =>', '')

  const { code } = transform(`
    <React.Fragment>
      ${jsx}
    </React.Fragment>
  `)

  const scope = { MDXTag, ...components, components, theme }
  const keys = Object.keys(scope)
  const values = keys.map(k => scope[k])

  const fn = new Function('React', ...keys, `return (${code})`)

  return fn(React, ...values)
}
