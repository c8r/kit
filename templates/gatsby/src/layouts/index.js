import React from 'react'
import PropTypes from 'prop-types'
import Helmet from 'react-helmet'
import { ThemeProvider } from 'styled-components'

import 'regenerator-runtime/runtime'

import theme from '../../theme'
import './index.css'

const Layout = ({ children, data }) => (
  <ThemeProvider theme={theme}>
    <div>
      <Helmet
        title={data.site.siteMetadata.title}
        meta={[
          { name: 'description', content: 'Sample' },
          { name: 'keywords', content: 'sample, something' }
        ]}
      />
      {children()}
    </div>
  </ThemeProvider>
)

Layout.propTypes = {
  children: PropTypes.func
}

export default Layout

export const query = graphql`
  query SiteTitleQuery {
    site {
      siteMetadata {
        title
      }
    }
  }
`
