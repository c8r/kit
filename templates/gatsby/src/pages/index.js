import React from 'react'
import Link from 'gatsby-link'

const IndexPage = () => (
  <div>
    <p>Welcome to your new Kit + Gatsby site.</p>
    <Link to="/kit/">Kit Library</Link>
  </div>
)

export default IndexPage
