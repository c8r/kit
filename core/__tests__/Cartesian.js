import React from 'react'
import styled from 'nano-style'
import { render } from 'react-testing-library'
import { color, space, fontSize } from 'styled-system'

import { Cartesian } from '../src'

const buttonProps = {
  children: 'Hello, world!',
  fontSize: [1, 2, 3],
  px: [2, 3],
  backgroundColor: ['pink', 'tomato', 'purple']
}

const Button = styled('a')(
  color,
  fontSize,
  space,
)

test('Cartesian renders all examples', () => {
  const { container } = render(<Cartesian {...buttonProps} component={Button} />)

  expect(container).toMatchSnapshot()
})
