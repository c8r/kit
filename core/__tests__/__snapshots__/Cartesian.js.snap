// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cartesian renders all examples 1`] = `
<div>
  <a
    backgroundcolor="pink"
    class="nano1kwchfs"
    font-size="1"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1kwchfs{font-size:14px}.nano1kwchfs{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nano1kwchfs"
    font-size="1"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1kwchfs{font-size:14px}.nano1kwchfs{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="purple"
    class="nano1kwchfs"
    font-size="1"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1kwchfs{font-size:14px}.nano1kwchfs{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="pink"
    class="nanoxw9r4l"
    font-size="1"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanoxw9r4l{font-size:14px}.nanoxw9r4l{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nanoxw9r4l"
    font-size="1"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanoxw9r4l{font-size:14px}.nanoxw9r4l{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="purple"
    class="nanoxw9r4l"
    font-size="1"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanoxw9r4l{font-size:14px}.nanoxw9r4l{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="pink"
    class="nano1cf2zn1"
    font-size="2"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1cf2zn1{font-size:16px}.nano1cf2zn1{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nano1cf2zn1"
    font-size="2"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1cf2zn1{font-size:16px}.nano1cf2zn1{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="purple"
    class="nano1cf2zn1"
    font-size="2"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nano1cf2zn1{font-size:16px}.nano1cf2zn1{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="pink"
    class="nanowuhht0"
    font-size="2"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanowuhht0{font-size:16px}.nanowuhht0{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nanowuhht0"
    font-size="2"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanowuhht0{font-size:16px}.nanowuhht0{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="purple"
    class="nanowuhht0"
    font-size="2"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nanowuhht0{font-size:16px}.nanowuhht0{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="pink"
    class="nanoxcuqpe"
    font-size="3"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nanoxcuqpe{font-size:20px}.nanoxcuqpe{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nanoxcuqpe"
    font-size="3"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nanoxcuqpe{font-size:20px}.nanoxcuqpe{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="purple"
    class="nanoxcuqpe"
    font-size="3"
    px="2"
  >
    Hello, world!
  </a>
  <style>
    .nanoxcuqpe{font-size:20px}.nanoxcuqpe{padding-left:8px;padding-right:8px}
  </style>
  <a
    backgroundcolor="pink"
    class="nano1ahjv9n"
    font-size="3"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nano1ahjv9n{font-size:20px}.nano1ahjv9n{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="tomato"
    class="nano1ahjv9n"
    font-size="3"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nano1ahjv9n{font-size:20px}.nano1ahjv9n{padding-left:16px;padding-right:16px}
  </style>
  <a
    backgroundcolor="purple"
    class="nano1ahjv9n"
    font-size="3"
    px="3"
  >
    Hello, world!
  </a>
  <style>
    .nano1ahjv9n{font-size:20px}.nano1ahjv9n{padding-left:16px;padding-right:16px}
  </style>
</div>
`;
